internal/controllers/project/actuator.go

```diff
@@ -0,0 +1,172 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package project
+
+import (
+	"context"
+	"iter"
+	"slices"
+
+	"github.com/gophercloud/gophercloud/v2/openstack/identity/v3/projects"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	orcv1alpha1 "github.com/k-orc/openstack-resource-controller/v2/api/v1alpha1"
+	generic "github.com/k-orc/openstack-resource-controller/v2/internal/controllers/generic/interfaces"
+	"github.com/k-orc/openstack-resource-controller/v2/internal/controllers/generic/progress"
+	orcerrors "github.com/k-orc/openstack-resource-controller/v2/internal/util/errors"
+	"github.com/k-orc/openstack-resource-controller/v2/internal/util/neutrontags"
+)
+
+// OpenStack resource types
+type (
+	osResourceT = projects.Project
+
+	createResourceActuator = generic.CreateResourceActuator[orcObjectPT, orcObjectT, filterT, osResourceT]
+	deleteResourceActuator = generic.DeleteResourceActuator[orcObjectPT, orcObjectT, osResourceT]
+	helperFactory          = generic.ResourceHelperFactory[orcObjectPT, orcObjectT, resourceSpecT, filterT, osResourceT]
+)
+
+type projectClient interface {
+	GetProject(context.Context, string) (*projects.Project, error)
+	ListProjects(context.Context, projects.ListOptsBuilder) iter.Seq2[*projects.Project, error]
+	CreateProject(context.Context, projects.CreateOptsBuilder) (*projects.Project, error)
+	DeleteProject(context.Context, string) error
+}
+
+type projectActuator struct {
+	osClient projectClient
+}
+
+var _ createResourceActuator = projectActuator{}
+var _ deleteResourceActuator = projectActuator{}
+
+func (projectActuator) GetResourceID(osResource *projects.Project) string {
+	return osResource.ID
+}
+
+func (actuator projectActuator) GetOSResourceByID(ctx context.Context, id string) (*projects.Project, progress.ReconcileStatus) {
+	project, err := actuator.osClient.GetProject(ctx, id)
+	if err != nil {
+		return nil, progress.WrapError(err)
+	}
+	return project, nil
+}
+
+func (actuator projectActuator) ListOSResourcesForAdoption(ctx context.Context, obj orcObjectPT) (iter.Seq2[*projects.Project, error], bool) {
+	if obj.Spec.Resource == nil {
+		return nil, false
+	}
+
+	listOpts := projects.ListOpts{
+		Name: string(getResourceName(obj)),
+		Tags: neutrontags.Join(obj.Spec.Resource.Tags),
+	}
+
+	return actuator.osClient.ListProjects(ctx, listOpts), true
+}
+
+func (actuator projectActuator) ListOSResourcesForImport(ctx context.Context, orcObject orcObjectPT, filter filterT) (iter.Seq2[*osResourceT, error], progress.ReconcileStatus) {
+	listOpts := projects.ListOpts{
+		Name: string(ptr.Deref(filter.Name, "")),
+		Tags: neutrontags.Join(filter.Tags),
+		// TagsAny:    neutrontags.Join(filter.TagsAny),
+		// NotTags:    neutrontags.Join(filter.NotTags),
+		// NotTagsAny: neutrontags.Join(filter.NotTagsAny),
```

Comment:
Keystone doesn't support these?

Reply:
It does, I have yet to clean up the PR. That will be part of the final PR.